added bands option to stacLayer

DanielJDufour opened the pull request:
This Pull Request adds the bands option to stacLayer.  You can use it to re-map the bands to [R, G, B] or [R,G,B,A] bands of the rendered image for assets that are passed to GeoRasterLayer.  As a part of this, it will automatically stretch the bands to fit in the image color space (0-255), so you don't have to worry about 16-bit images.  The stretching uses the statistics that are dynamically calculated by GeoRasterLayer based on rendered pixels.

Example:
```js
// only visualize second band (index 1)
stacLayer(data, { bands: [1, 1, 1] })

// in image with [red, green, blue, nir] use [nir, green, blue]
stacLayer(data, { bands: [3, 1, 2] })
```

Open to thoughts and suggestions.

This PR is indeed hacky, requiring a few features of GeoRasterLayer that are not publicly documented.  This functionality might be added to GeoRasterLayer at some point, which would simplify the stac-layer code.

m-mohr commented:
This looks pretty good to me. The only place where I'm not sure whether the statistics/stretching make sense is the alpha/transparency layer. It might make more sense that for alpha (if provided) the range for integers is fixed to the range of the data type and for floats it is fixed to 0-1 maybe?

DanielJDufour commented:
@m-mohr , I'll try to find/create some examples, so we can test/try out the different options ... and I can provide a good answer to your very good question :-)

m-mohr commented:
Any news? :-)

DanielJDufour commented:
I did a little more thinking about this (and I'm not sure why I didn't think of this earlier), but I believe GeoTIFFs should have the alpha channel marked as such in the metadata via https://www.awaresystems.be/imaging/tiff/tifftags/extrasamples.html.  I still think incorporating some of the logic you mentioned above as a double safety check would be good too.

DanielJDufour commented:
I think I found a file that I can write a test with: https://github.com/GeoTIFF/test-data/blob/main/files/gadas.tif.  It comes from a PNG export from an ArcGIS Server, converted to a GeoTIFF.

Here's the relevant metadata from geotiff.js:
```js
BitsPerSample: Uint16Array(4) [ 8, 8, 8, 8 ],
ExtraSamples: Uint16Array(1) [ 2 ], // 2 means "Unassociated alpha channel"
```

DanielJDufour commented:
Hey, @m-mohr .  I believe this is ready for your review again.   A few notes:
- I added https://github.com/geotiff/geotiff-epsg-code to better support data produced by ESRI software (and work with a wider range of random test data, specifically the test tile with an alpha band)
- Added a setup script that pulls in https://github.com/GeoTIFF/test-data/ (geotiff.js also uses some of this test-data)
- Added in alpha band detection and rendering rules you expressed above (0-1) and (0-100).  I also added in an if/rule for a float that is 0-255 (indicating the dataset used to be uint8, but was converted to float for some reason), and no range detection where min == max.  

I also added a test file that allows you to change the bands mapping via a url param.

Let me know if you have any questions or concerns :-)

It's always tough trying to decide how to automatically visualize random datasets, so definitely more help the better :-)

m-mohr commented on README.md:
```diff
@@ -44,6 +44,7 @@ The following options are supported:
 
 | **Option**                   | **Data Type** | **Default value**                          | **Description** |
 | ---------------------------- | ------------- | ------------------------------------------ |-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
+| `bands`                      | array of numbers | undefined | an array mapping the bands to the output bands. must be of length 3 or 4. |
```
I'd suggest making the one-band use case a bit simpler by allowing 1 and 2 elements, too.
- 1 element means single band w/o alpha (i.e. expand from e.g. [0] to [0,0,0])
- 2 elements mean single band with alpha (i.e. expand from e.g. [0, 1] to [0,0,0,1])
- 3 and 4 as defined below.

```suggestion
| `bands`                      | array of numbers | undefined | an array mapping the bands to the output bands. must be of length 1, 2, 3 or 4. |
```

m-mohr replied:
=> #57

DanielJDufour replied:
That's a great idea!

m-mohr reviewed (approved):
Looks good to me, thanks a lot!

I've added one comment, but maybe we can just make that into an issue and merge this anyway. (I want to work on the stac-js integration, but not run into heave merge conflicts ;-) )
